### PR TITLE
Comment input field as almEditable element, fixes #722

### LINKED ISSUE #722: Make WI comment section almEditable
Replace regular text area with `<div contenteditable ... >` using almEditable directive for consistency.

Part of userstory: #651

### PR DESCRIPTION
**Pull request by debloper**
- Replaces textarea with contenteditable div, by almEditable directive
- Fixes an index counter issue introduced by #731, that was breaking tests
- Workflow fixes for the new element (content reset on various actions)
- Optimized the way insertion into comment collection for a WI was being handled

Signed-off-by: Soumya Deb <debloper@gmail.com>

### CODE REVIEW

**Comment by vikram-raj** [comment]
Following behavior I noticed when I open the detail page for the first time after page reload:
1. Comment Box is automatically focused. Do we need this behavior?
2. "Add a new comment..." is not a placeholder of the comment box. "Add a new comment..." should be a placeholder.

File: src/app/work-item/work-item-detail/work-item-comment/work-item-comment.component.scss [hunk 1]
```diff
@@ -9,6 +9,9 @@
     > div{
         padding: 0;
     }
+    #wi-comment-add-comment {
+        height: em(64);
```
Comment (nimishamukherjee) [inline src/app/work-item/work-item-detail/work-item-comment/work-item-comment.component.scss]: We need scroll on the almEditable to support multi-line comments.
On demo.almighty
ondemo

Your PR
commentsoverflow

File: src/app/work-item/work-item-detail/work-item-comment/work-item-comment.component.html [hunk 2]
```diff
@@ -3,12 +3,12 @@
...
         <div class="col-md-12 col-sm-12 col-xs-12" *ngIf="true">
             <div *ngIf="loggedIn" class="">
-                <textarea rows="2" type="text" id="wi-comment-add-comment" class="form-control width100" [(ngModel)]="comment.attributes.body"
-                    [style.height]="descHeight" [style.resize]="descResize" (keydown.enter)="createComment($event)" (keyup.enter)="preventDef($event)"
-                    placeholder="Add a comment here...">
-                </textarea>
+                <div almEditable id="wi-comment-add-comment" class="form-control width100"
+                    (focus)="resetCommentDraft($event)" (blur)="resetCommentDraft($event)"
+                    (keydown.enter)="createComment($event)" (keyup.enter)="preventDef($event)"
+                    [textContent]="'Add a new comment...'" #comment></div>
```
Comment (nimishamukherjee) [inline src/app/work-item/work-item-detail/work-item-comment/work-item-comment.component.html]: "Add a comment" should be placeholder.

Comment (sanbornsen) [inline src/app/work-item/work-item-detail/work-item-comment/work-item-comment.component.html]: I agree that there is no placeholder in div, but it should act like one.

**Comment by sanbornsen** [comment]
- The comment field should not be focused by default.
-  Focusing on the comment field should remove the place holder and make it empty
- Place holder should have a different text style as it's done in description
- Use <p> tag instead of <div> for auto height adjust and don't specify any height for the element anywhere.

File: src/app/work-item/work-item-detail/work-item-comment/work-item-comment.component.html (hunk 2 above)
Comment (debloper) [inline src/app/work-item/work-item-detail/work-item-comment/work-item-comment.component.html]: contenteditable div doesn't have placeholders, but we can add styles etc afterwards to make it look like placeholder. Considering extending almEditable directive for that purpose.

For now the resetCommentDraft handles it such a way the placeholder can't be posted as comment (apart from the catch by @sanbornsen about when it's focused by default).

File: src/app/work-item/work-item-detail/work-item-comment/work-item-comment.component.scss [hunk 3]
```diff
@@ -9,6 +9,19 @@
...
+    #wi-comment-add-comment {
+        /* Overrides pf's form-control style's height property */
+        /* The min-height is needed to avoid zero-height element */
+        /* The min-height is same as form-control's default height */
+        height: auto;
+        min-height: 26px;
+
+        /* Helper class, should be moved out when almEditable is extended to support placeholders */
+        &.placeholder {
+            color: #aaa;
```
Comment (debloper) [inline src/app/work-item/work-item-detail/work-item-comment/work-item-comment.component.scss]: Good idea about using color variables.
I'll go with `color-pf-black-400` (we don't need white here).

**Comment by vikram-raj** [comment]
@debloper is your tab size is 4?  Can you change it to 2? And it is good if you install EditorConfig for VS Code.

**Comment by debloper** [comment]
I just went with the consistency of the particular files.
My own settings are at 2 space, but I don't wanna mix it up in a file where the whole file is 4 space indented.

File: src/app/work-item/work-item-detail/work-item-comment/work-item-comment.component.scss [hunk 4]
```diff
@@ -9,6 +9,19 @@
     > div{
         padding: 0;
     }
+    #wi-comment-add-comment {
+        /* Overrides pf's form-control style's height property */
+        /* The min-height is needed to avoid zero-height element */
+        /* The min-height is same as form-control's default height */
+        height: auto;
+        min-height: 26px;
```
Comment (nimishamukherjee) [inline src/app/work-item/work-item-detail/work-item-comment/work-item-comment.component.scss]: @SMahil can you confirm if min-height needs to be in em or px?

File: src/app/work-item/work-item-detail/work-item-comment/work-item-comment.component.html [hunk 5]
```diff
@@ -3,12 +3,12 @@
         <!-- Refactoring needed to toogle view -->
         <div class="col-md-12 col-sm-12 col-xs-12" *ngIf="true">
             <div *ngIf="loggedIn" class="">
-                <textarea rows="2" type="text" id="wi-comment-add-comment" class="form-control width100" [(ngModel)]="comment.attributes.body"
-                    [style.height]="descHeight" [style.resize]="descResize" (keydown.enter)="createComment($event)" (keyup.enter)="preventDef($event)"
-                    placeholder="Add a comment here...">
-                </textarea>
+                <p almEditable id="wi-comment-add-comment" class="form-control placeholder width100"
+                    (focus)="resetCommentDraft($event)" (blur)="resetCommentDraft($event)"
```
Comment (nimishamukherjee) [inline src/app/work-item/work-item-detail/work-item-comment/work-item-comment.component.html]: @debloper can you confirm with UX team (screenshot or video) on the comment box being focused when the detail's panel opens. I'm not convinced that we should be doing that :)

**Comment by nimishamukherjee** [comment]
@debloper unit tests are failing locally.

File: src/app/work-item/work-item-detail/work-item-comment/work-item-comment.component.scss (hunk 4 above)
Comment (debloper) [inline src/app/work-item/work-item-detail/work-item-comment/work-item-comment.component.scss]: I went with the BS/PF default for text-input (and they go with px).

Need to check with the UXD on this. With textarea, the field used to be 3-rows in height, not sure that was by default or as per UXD definition.

File: src/app/work-item/work-item-detail/work-item-comment/work-item-comment.component.html (hunk 5 above)
Comment (debloper) [inline src/app/work-item/work-item-detail/work-item-comment/work-item-comment.component.html]: Alright, this needs fixing.

**Comment by SMahil** [comment]
@nimishamukherjee @debloper to keep height in em or px We will ask UX Team and i think we should do this change in separate PR as this will going to change several places

**Comment by debloper** [comment]
@SMahil thanks, sounds good.

File: src/app/work-item/work-item-detail/work-item-comment/work-item-comment.component.html (hunk 5 above)
Comment (debloper) [inline src/app/work-item/work-item-detail/work-item-comment/work-item-comment.component.html]: @nimishamukherjee can you verify?
I have already had fixed in in previous push, and rechecked again. Doesn't happen for me.

**Comment by nimishamukherjee** [comment]
[test]

**Pull request merged**